### PR TITLE
Duplex s2s new speech decoder

### PR DESCRIPTION
**Pull request by cchen1436**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

The implementation of duplex s2s model with an independent speech decoder, which surports FSDP training and AR decoding.

**Collection**: [duplex_s2s]

# Changelog

- Support AR decoding for both text and speech.
- Support FSDP training

# Usage

The usage doc can be found at: https://docs.google.com/document/d/17z9JeCK8Ht-ojLp1H8ynJqr8Pz7aGdohy0C0PYIvXvY/edit?usp=sharing

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: nemo/collections/duplex_s2s/models/duplex_s2s_model.py [hunk 1]
```diff
@@ -524,28 +528,6 @@ def configure_optimizers(self):
             ans["lr_scheduler"] = {"scheduler": lr_scheduler, "interval": "step", "frequency": 1}
         return ans
 
-    @property
-    def oomptimizer_schema(self) -> dict:
```
Comment (pzelasko) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model.py:528]: Please keep this.

File: nemo/collections/duplex_s2s/models/duplex_s2s_model.py [hunk 2]
```diff
@@ -556,23 +538,6 @@ def configure_model(self) -> None:
         if (tp_mesh := device_mesh["tensor_parallel"]).size() > 1:
             self._use_tp = True
 
-            # TODO: Distributing embeddings with TP in this setup is tricky
```
Comment (pzelasko) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model.py:559]: Please keep this for now, we should sort it out to get max benefit from TP .

File: nemo/collections/duplex_s2s/models/duplex_s2s_model.py [hunk 3]
```diff
@@ -599,8 +564,6 @@ def configure_model(self) -> None:
                     "mlp.gate_proj": ColwiseParallel(),
                     "mlp.up_proj": ColwiseParallel(),
                     "mlp.down_proj": RowwiseParallel(output_layouts=Shard(1)),
-                    # "pre_feedforward_layernorm": SequenceParallel(),
-                    # "post_feedforward_layernorm": SequenceParallel(),
```
Comment (pzelasko) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model.py:603]: Please keep this for awareness for now, some HF models have these extra layers, we'll need a more generic implementation here to support this.

File: nemo/collections/duplex_s2s/models/duplex_s2s_model.py [hunk 4]
```diff
@@ -654,6 +617,424 @@ def configure_model(self) -> None:
...
+                input_embeds = torch.cat([input_embeds, pad], dim=1)
+        else:
+            T = T_local
+
+        # Apply channel weight
+        input_embeds *= self.cfg.get("duplex_user_channel_weight", 1.0)
+
+        # This cache is for self.llm
+        cache = DynamicCache()
+        # Call reset_input_and_kv_cache to enable cache for TransformerARSpeechDecoder
```
Comment (pzelasko) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model.py:899]: it would be good if TransformerARSpeechDecoder had explicit Cache argument that can be managed by the user.

File: nemo/collections/duplex_s2s/modules/speech_generation.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                audio_embeddings.append(nn.Embedding(num_audio_tokens_per_codebook, self.speech_decoder_parms["d_model"]))
+
+            self.audio_embeddings = nn.ModuleList(audio_embeddings)
+
+    def forward(self, hidden_states, speech_mask, input_audio_tokens=None, return_raw_logits=False):
+        # Megatron LLM parallel training returns T, B, F so reshape it
+        # T, B, F = hidden_states.size()
+        hidden_states = hidden_states.transpose(0, 1).contiguous() # .reshape(B, T, F) # from [T, B, F] to [B, T, F]
+        # input cache needed due our transformer kv cache implementation expect the whole left context
+        if self.use_input_cache:
```
Comment (pzelasko) [inline nemo/collections/duplex_s2s/modules/speech_generation.py:47]: Can we transform stateful KV cache to argument+return value for a pure functional API?

File: nemo/collections/tts/models/audio_codec.py [hunk 6]
```diff
@@ -32,6 +32,7 @@
     SISDRLoss,
```
Comment (pzelasko) [inline nemo/collections/tts/models/audio_codec.py:32]: Will the changes in TTS collection be merged in a separate PR? In that case we should drop them here (CC @zhehuaichen @Edresson @blisc)

**Review by pzelasko — CHANGES_REQUESTED** [review]
Regarding `DuplexS2SModelSpeechDecoder` structure, here are my suggestions.

Option 1.
* Move `DuplexS2SModelSpeechDecoder` to a separate file.
* Instead of inheriting from `DuplexS2SModel`, let's have `self.s2s = DuplexS2SModel(...)`.
* Some methods such as `prepare_inputs`, `configure_model` can call `super().<corresponding-method>` and then add/modify the result.

Option 2.
* Copy entire `duplex_s2s_model.py` to a separate file.
* Remove inheritance and apply all changes from `DuplexS2SModelSpeechDecoder` to the base class in the new file, and then rename it to `DuplexS2SModelSpeechDecoder`.
* We can extract some common parts (configure model, validation, metrics) to separate utilities

Option 3.
* Keep the inheritance as it is.

I'm afraid that option 3 will make us suffer in the longer run, degrading the codebase to the same state as the old `speech_llm` collection over time. If we choose that, we should limit ourselves to at most 1 level of inheritance to minimize this risk (i.e., `DuplexS2SModelSpeechDecoder` can't be inherited from) - but I'm not sure how that's better than option 2.

Option 1 would be the best IMO, but I'm not 100% sure if composition will be sufficient (e.g., if some internal behavior of DuplexS2SModel needs to be modified).

**Comment by kevinhu-nv** [comment]
To balance implementation complexity, I think 2 may be a good option. 1) may increase the complexity to achieve parity. I think for the SDV2 version we already overwrite a lot of methods so may be OK to just split the whole class out from DuplexS2SModel. 3) has the risk of too many levels of inheritance - let's avoid that if not much time is needed.

**Comment by Edresson** [comment]
I agree with Kevin, I think option 2 would be a better approach.

**Comment by zhehuaichen** [comment]
+1 and let's do `We can extract some common parts (configure model, validation, metrics) to separate utilities` as much as we can while treating sdv2 as the primary class to be clean so as to be carried over

File: low-frame-rate-codec.patch [hunk 7]
```diff
@@ -0,0 +1,1265 @@
+diff --git a/nemo/collections/tts/models/audio_codec.py b/nemo/collections/tts/models/audio_codec.py
```
Comment (pzelasko) [inline low-frame-rate-codec.patch:1]: This file needs to be removed from the PR

File: nemo/collections/duplex_s2s/modules/speech_generation.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+                speech_decoder_input = torch.cat([speech_decoder_input, speech_decoder_input_zeros], dim=0)
+                # duplicate mask to match the new shape
+                speech_mask = torch.cat([speech_mask, speech_mask], dim=0)
+                # if cond on prev tokens enabled, so duplicate the tokens to the new shape
+                if self.cond_on_prev_audio_tokens:
+                    input_audio_tokens = torch.cat([input_audio_tokens, input_audio_tokens], dim=0)
+
+        if self.cond_on_prev_audio_tokens:
+            if self.detach_input:
+                input_audio_tokens = input_audio_tokens.detach()
```
Comment (zhehuaichen) [inline nemo/collections/duplex_s2s/modules/speech_generation.py:98]: we should actually remove this line after confirming with Edresson. We only need to detach hidden states from the llm but not the audio tokens from auto-regressive

File: nemo/collections/duplex_s2s/modules/speech_generation.py (hunk 5 above)
Comment (pzelasko) [reply nemo/collections/duplex_s2s/modules/speech_generation.py:47]: Deferred for follow up work

File: nemo/collections/tts/models/audio_codec.py (hunk 6 above)
Comment (Edresson) [reply nemo/collections/tts/models/audio_codec.py:37]: I talked with @blisc, and we think merging these changes here is fine. I also updated the codec classes to match those in the old codebase.

File: low-frame-rate-codec.patch (hunk 7 above)
Comment (Edresson) [reply low-frame-rate-codec.patch]: Done

File: nemo/collections/duplex_s2s/modules/speech_generation.py (hunk 8 above)
Comment (Edresson) [reply nemo/collections/duplex_s2s/modules/speech_generation.py]: I removed it.

File: examples/duplex_s2s/s2s_duplex_train.py [hunk 9]
```diff
@@ -19,7 +19,7 @@
 from omegaconf import OmegaConf
 
 from nemo.collections.duplex_s2s.data.datamodule import S2SDataModule
```
Comment (Edresson) [inline examples/duplex_s2s/s2s_duplex_train.py:21]: I think would be nice if we could create a s2s_duplex_speech_decoder_train.py and keep s2s_duplex_train.py file for INTERSPEECH-like architecture.

File: nemo/collections/duplex_s2s/models/duplex_s2s_model.py [hunk 10]
```diff
@@ -18,16 +18,17 @@
 from itertools import chain
 from pathlib import Path
 from typing import Any
-
+from lightning.pytorch.utilities import rank_zero_only
```
Comment (Edresson) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model.py:21]: Remove this unused import.

File: nemo/collections/duplex_s2s/models/duplex_s2s_model.py [hunk 11]
```diff
@@ -18,16 +18,17 @@
...
 import sacrebleu
 import torch
 from lightning import LightningModule
 from omegaconf import open_dict
 from torch import Tensor
 from torch.distributed import init_device_mesh
+import torch.distributed as dist
 from torch.distributed.algorithms._checkpoint.checkpoint_wrapper import checkpoint_wrapper
-from torch.distributed.fsdp import MixedPrecisionPolicy, fully_shard
+from torch.distributed.fsdp import MixedPrecisionPolicy, fully_shard, FullyShardedDataParallel
```
Comment (Edresson) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model.py:31]: Remove this unused import.

File: nemo/collections/duplex_s2s/models/duplex_s2s_model.py [hunk 12]
```diff
@@ -44,12 +45,14 @@
 from nemo.collections.common.parts.optional_cuda_graphs import WithOptionalCudaGraphs
 from nemo.collections.common.tokenizers import AutoTokenizer
 from nemo.collections.duplex_s2s.modules import AudioPerceptionModule
+from nemo.collections.duplex_s2s.modules import TransformerARSpeechDecoder
```
Comment (Edresson) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model.py:48]: Remove this unused import.

File: nemo/collections/duplex_s2s/models/duplex_s2s_model.py [hunk 13]
```diff
@@ -18,16 +18,17 @@
...
-
+from lightning.pytorch.utilities import rank_zero_only
 import hydra
 import sacrebleu
 import torch
 from lightning import LightningModule
 from omegaconf import open_dict
 from torch import Tensor
 from torch.distributed import init_device_mesh
+import torch.distributed as dist
```
Comment (Edresson) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model.py:29]: Remove this unused import.

File: nemo/collections/duplex_s2s/models/duplex_s2s_model.py [hunk 14]
```diff
@@ -44,12 +45,14 @@
 from nemo.collections.common.parts.optional_cuda_graphs import WithOptionalCudaGraphs
 from nemo.collections.common.tokenizers import AutoTokenizer
 from nemo.collections.duplex_s2s.modules import AudioPerceptionModule
+from nemo.collections.duplex_s2s.modules import TransformerARSpeechDecoder
 from nemo.collections.tts.models import AudioCodecModel
 from nemo.core.neural_types import AudioSignal, LabelsType, LengthsType, NeuralType
 from nemo.utils import logging
 
+import torch.distributed as dist
```
Comment (Edresson) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model.py:53]: Remove this unused import.

File: nemo/collections/duplex_s2s/models/duplex_s2s_model_speech_decoder.py [hunk 15]
```diff
@@ -0,0 +1,1083 @@
...
+from nemo.collections.duplex_s2s.modules import AudioPerceptionModule
+from nemo.collections.duplex_s2s.modules import TransformerARSpeechDecoder
+from nemo.collections.tts.models import AudioCodecModel
+from nemo.core.neural_types import AudioSignal, LabelsType, LengthsType, NeuralType
+from nemo.utils import logging
+
+import torch.distributed as dist
+
+
+class DuplexS2SModel(LightningModule):
```
Comment (Edresson) [inline nemo/collections/duplex_s2s/models/duplex_s2s_model_speech_decoder.py:56]: I think we should remove this class and add its methods to the DuplexS2SModelSpeechDecoder class.
